File: python-cleanroom/packages/botas/src/bot_application.py
from __future__ import annotations
import json
import os
import time
import base64
import re
from typing import Any, Awaitable, Callable, Optional
from typing_extensions import Protocol, TypeVar

import httpx
import jwt

from .models import CoreActivity, ResourceResponse, InvokeResponse, ChannelAccount, Conversation
from .turn_context import TurnContext


class _BotApplicationBot:
    """BotApplication instance returned by on() for chaining."""
    
    def __init__(self, bot: BotApplication) -> None:
        self._bot = bot
    
    def on(self, activity_type: str, handler):
        self._bot.on(activity_type, handler)
        return self
    
    def use(self, middleware):
        self._bot.use(middleware)
        return self


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .turn_context import TurnContext


T = TypeVar("T")
NextTurn = Callable[[], Awaitable[None]]


class TurnMiddleware(Protocol):
    async def on_turn(self, context: TurnContext, next: NextTurn) -> None: ...


P = TypeVar("P", bound=TurnMiddleware)


class BotApplication:
    def __init__(
        self,
        options: Optional[BotApplicationOptions] = None,
    ) -> None:
        self._handlers: dict[str, Callable[[TurnContext], Awaitable[None]]] = {}
        self._middleware: list[TurnMiddleware] = []
        self._on_activity: Optional[Callable[[TurnContext], Awaitable[None]]] = None
        self._http_client = httpx.AsyncClient()
        self._outbound_token = ""
        self._token_expiry = 0
        opts = options or BotApplicationOptions()
        self._client_id = opts.client_id or os.getenv("CLIENT_ID", "")
        self._client_secret = opts.client_secret or os.getenv("CLIENT_SECRET", "")
        self._tenant_id = opts.tenant_id or os.getenv("TENANT_ID", "common")

    def on(
        self, activity_type: str, handler: Callable[[TurnContext], Awaitable[None]]
    ) -> "BotApplication":
        self._handlers[activity_type.lower()] = handler
        return self

    @property
    def on_activity(self) -> Optional[Callable[[TurnContext], Awaitable[None]]]:
        return self._on_activity

    @on_activity.setter
    def on_activity(
        self, value: Optional[Callable[[TurnContext], Awaitable[None]]]
    ) -> None:
        self._on_activity = value

    def use(self, middleware: TurnMiddleware) -> "BotApplication":
        self._middleware.append(middleware)
        return self

    async def process_body(self, body: str) -> InvokeResponse | None:
        activity = CoreActivity(**json.loads(body))
        return await self._process_activity(activity)

    async def send_activity_async(
        self,
        service_url: str,
        conversation_id: str,
        activity: CoreActivity,
    ) -> ResourceResponse | None:
        if not service_url or not conversation_id:
            return None

        url = (
            self._normalize_service_url(service_url)
            + "v3/conversations/"
            + self._truncate_conversation_id(conversation_id)
            + "/activities"
        )

        token = await self._get_outbound_token()

        response = await self._http_client.post(
            url,
            json=activity.model_dump(by_alias=True, exclude_none=True),
            headers={"Authorization": f"Bearer {token}"},
        )
        response.raise_for_status()
        return ResourceResponse(**response.json())

    async def _process_activity(
        self, activity: CoreActivity
    ) -> InvokeResponse | None:
        if not self._validate_required_fields(activity):
            return InvokeResponse(status="error", body={"error": "invalid activity"})
        if not self._validate_service_url(activity.service_url):
            return InvokeResponse(status="error", body={"error": "invalid serviceUrl"})

        turn_context = TurnContext(self, activity)
        final_handler = lambda: self._dispatch(turn_context)

        await self._run_middleware_pipeline(turn_context, final_handler)
        return InvokeResponse(status="ok")

    async def _run_middleware_pipeline(
        self, context: TurnContext, final_handler: NextTurn
    ) -> None:
        async def run():
            await final_handler()
            return None

        next: NextTurn = run
        for middleware in reversed(self._middleware):
            current_next = next

            async def wrapper(__next: NextTurn, ctx=context, mw=middleware) -> None:
                await mw.on_turn(ctx, __next)

            next = wrapper

        await next()

    async def _dispatch(self, turn_context: TurnContext) -> None:
        handler = self._on_activity
        if not handler:
            handler = self._handlers.get(turn_context.activity.type.lower())  # type: ignore

        if handler:
            try:
                await handler(turn_context)
            except Exception as e:
                raise BotHandlerError(e, turn_context.activity)

    async def _get_outbound_token(self) -> str:
        if self._outbound_token and time.time() < self._token_expiry - 300:
            return self._outbound_token

        token_url = (
            f"https://login.microsoftonline.com/{self._tenant_id}/oauth2/v2.0/token"
        )

        response = await self._http_client.post(
            token_url,
            data={
                "grant_type": "client_credentials",
                "client_id": self._client_id,
                "client_secret": self._client_secret,
                "scope": "https://api.botframework.com/.default",
            },
        )
        result = response.json()
        self._outbound_token = result["access_token"]
        self._token_expiry = time.time() + result["expires_in"]
        return self._outbound_token

    def _validate_token(self, token: str) -> bool:
        try:
            parts = token.split(".")
            if len(parts) != 3:
                return False

            payload = json.loads(
                base64.urlsafe_b64decode(parts[1] + "==").decode()
            )

            aud = payload.get("aud")
            iss = payload.get("iss")
            tid = payload.get("tid")
            exp = payload.get("exp")

            if exp < time.time():
                return False

            valid_audiences = [
                self._client_id,
                f"api://{self._client_id}",
                "https://api.botframework.com",
            ]
            if aud not in valid_audiences:
                return False

            valid_issuers = [
                "https://api.botframework.com",
                f"https://sts.windows.net/{tid}/",
                f"https://login.microsoftonline.com/{tid}/v2",
                f"https://login.microsoftonline.com/{tid}/v2.0",
            ]
            if iss not in valid_issuers:
                return False

            return True
        except Exception:
            return False

    def _validate_required_fields(self, activity: CoreActivity) -> bool:
        return bool(activity.type and activity.service_url and activity.conversation.id)

    def _validate_service_url(self, service_url: str) -> bool:
        try:
            from urllib.parse import urlparse

            parsed = urlparse(service_url)
            host = parsed.netloc.lower()

            if host in ("localhost", "127.0.0.1"):
                return parsed.scheme in ("http", "https")

            allowed_patterns = [
                "botframework.com",
                "botframework.us",
                "botframework.cn",
                "trafficmanager.net",
            ]

            return any(host.endswith(p) for p in allowed_patterns) and parsed.scheme == "https"
        except Exception:
            return False

    def _normalize_service_url(self, service_url: str) -> str:
        service_url = service_url.rstrip("/")
        if not service_url.endswith("/"):
            service_url += "/"
        return service_url

    def _truncate_conversation_id(self, conversation_id: str) -> str:
        semicolon_index = conversation_id.find(";")
        if semicolon_index > 0:
            conversation_id = conversation_id[:semicolon_index]
        return conversation_id

    async def close(self) -> None:
        await self._http_client.aclose()


class BotApplicationOptions:
    def __init__(
        self,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        tenant_id: Optional[str] = None,
        managed_identity_client_id: Optional[str] = None,
        token_factory: Optional[Callable[[str, str], Awaitable[str]]] = None,
    ) -> None:
        self.client_id = client_id
        self.client_secret = client_secret
        self.tenant_id = tenant_id
        self.managed_identity_client_id = managed_identity_client_id
        self.token_factory = token_factory


class BotHandlerError(Exception):
    def __init__(self, cause: Exception, activity: CoreActivity) -> None:
        super().__init__("Handler error")
        self.cause = cause
        self.activity = activity

File: node-cleanroom/packages/botas/src/bot-application.ts
import type { IncomingMessage, ServerResponse } from 'http';
import jwt from 'jsonwebtoken';
import jwksClient from 'jwks-rsa';
import type { CoreActivity, ResourceResponse, InvokeResponse } from './core-activity.js';
import type { TurnContext } from './turn-context.js';
import { TurnContext as TurnContextClass } from './turn-context.js';

export type NextTurn = () => Promise<void>;

export type TurnMiddleware = (
  context: TurnContext,
  next: NextTurn
) => Promise<void>;

interface BotApplicationOptions {
  clientId?: string;
  clientSecret?: string;
  tenantId?: string;
  managedIdentityClientId?: string;
  token?: (scope: string, tenantId: string) => Promise<string>;
}

export class BotApplication {
  private readonly handlers = new Map<string, (ctx: TurnContext) => Promise<void>>();
  private readonly middleware: TurnMiddleware[] = [];
  private onActivityHandler?: (ctx: TurnContext) => Promise<void>;
  private readonly httpClient: globalThis.fetch;
  private readonly options: Required<BotApplicationOptions>;
  private outboundToken = '';
  private tokenExpiry = 0;

  constructor(options: BotApplicationOptions = {}) {
    this.options = {
      clientId: options.clientId ?? process.env.CLIENT_ID ?? '',
      clientSecret: options.clientSecret ?? process.env.CLIENT_SECRET ?? '',
      tenantId: options.tenantId ?? process.env.TENANT_ID ?? 'common',
      managedIdentityClientId: options.managedIdentityClientId ?? process.env.MANAGED_IDENTITY_CLIENT_ID ?? '',
      token: options.token ?? (() => Promise.resolve('')),
    };
    this.httpClient = globalThis.fetch;
  }

  on(type: string, handler: (ctx: TurnContext) => Promise<void>): this {
    this.handlers.set(type.toLowerCase(), handler);
    return this;
  }

  get onActivity(): ((ctx: TurnContext) => Promise<void>) | undefined {
    return this.onActivityHandler;
  }

  set onActivity(value: ((ctx: TurnContext) => Promise<void>) | undefined) {
    this.onActivityHandler = value;
  }

  use(middleware: TurnMiddleware): this {
    this.middleware.push(middleware);
    return this;
  }

  async processAsync(req: IncomingMessage, res: ServerResponse): Promise<void> {
    if (this.options.clientId) {
      const authHeader = req.headers.authorization;
      if (!authHeader?.startsWith('Bearer ')) {
        res.statusCode = 401;
        res.end();
        return;
      }

      const token = authHeader.slice(7);
      const valid = await this.validateToken(token);
      if (!valid) {
        res.statusCode = 401;
        res.end();
        return;
      }
    }

    await this.processActivityInternal(req, res);
  }

  async processBody(body: string): Promise<InvokeResponse | undefined> {
    const activity = JSON.parse(body) as CoreActivity;
    return this.processActivityObject(activity);
  }

  private async processActivityInternal(
    req: IncomingMessage,
    res: ServerResponse
  ): Promise<void> {
    const chunks: Buffer[] = [];
    for await (const chunk of req) {
      chunks.push(chunk);
    }
    const body = Buffer.concat(chunks).toString();

    await this.processActivityObject(JSON.parse(body), res);
  }

  private async processActivityObject(
    activity: CoreActivity,
    res?: ServerResponse
  ): Promise<InvokeResponse | undefined> {
    if (!this.validateRequiredFields(activity)) {
      if (res) {
        res.statusCode = 400;
        res.setHeader('Content-Type', 'application/json');
        res.end('{}');
      }
      return undefined;
    }

    if (!this.validateServiceUrl(activity.serviceUrl)) {
      if (res) {
        res.statusCode = 400;
        res.setHeader('Content-Type', 'application/json');
        res.end('{}');
      }
      return undefined;
    }

    const turnContext = new TurnContextClass(this, activity);
    const finalHandler = async () => {
      await this.dispatch(turnContext);
    };

    await this.runMiddlewarePipeline(turnContext, finalHandler);

    if (res) {
      res.statusCode = 200;
      res.setHeader('Content-Type', 'application/json');
      res.end('{}');
    }

    return undefined;
  }

  private async runMiddlewarePipeline(
    context: TurnContext,
    finalHandler: NextTurn
  ): Promise<void> {
    let next = finalHandler;
    for (let i = this.middleware.length - 1; i >= 0; i--) {
      const middleware = this.middleware[i];
      const currentNext = next;
      next = async () => {
        await middleware(context, currentNext);
      };
    }
    await next();
  }

  private async dispatch(context: TurnContext): Promise<void> {
    let handler = this.onActivityHandler;

    if (!handler) {
      handler = this.handlers.get(context.activity.type.toLowerCase());
    }

    if (handler) {
      try {
        await handler(context);
      } catch (err) {
        throw new BotHandlerError(err as Error, context.activity);
      }
    }
  }

  async sendActivityAsync(
    serviceUrl: string,
    conversationId: string,
    activity: Partial<CoreActivity>
  ): Promise<ResourceResponse | undefined> {
    if (!serviceUrl || !conversationId) {
      return undefined;
    }

    const url = this.normalizeServiceUrl(serviceUrl) + 'v3/conversations/' +
      this.truncateConversationId(conversationId) + '/activities';

    const token = await this.getOutboundToken();

    const response = await this.httpClient(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(activity),
    });

    if (!response.ok) {
      throw new Error(`Failed to send activity: ${response.status}`);
    }

    return response.json() as Promise<ResourceResponse>;
  }

  private async getOutboundToken(): Promise<string> {
    if (this.outboundToken && Date.now() < this.tokenExpiry - 300000) {
      return this.outboundToken;
    }

    const tokenUrl = `https://login.microsoftonline.com/${this.options.tenantId}/oauth2/v2.0/token`;

    const params = new URLSearchParams({
      grant_type: 'client_credentials',
      client_id: this.options.clientId,
      client_secret: this.options.clientSecret,
      scope: 'https://api.botframework.com/.default',
    });

    const response = await this.httpClient(tokenUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    const result = await response.json() as {
      access_token: string;
      expires_in: number;
    };

    this.outboundToken = result.access_token;
    this.tokenExpiry = Date.now() + result.expires_in * 1000;

    return this.outboundToken;
  }

  private async validateToken(token: string): Promise<boolean> {
    try {
      const parts = token.split('.');
      if (parts.length !== 3) return false;

      const payload = JSON.parse(
        Buffer.from(parts[1], 'base64').toString()
      );

      const aud = payload.aud as string;
      const iss = payload.iss as string;
      const tid = payload.tid as string;
      const exp = payload.exp as number;

      if (exp < Date.now() / 1000) return false;

      const validAudiences = [
        this.options.clientId,
        `api://${this.options.clientId}`,
        'https://api.botframework.com',
      ];
      if (!validAudiences.includes(aud)) return false;

      const validIssuers = [
        'https://api.botframework.com',
        `https://sts.windows.net/${tid}/`,
        `https://login.microsoftonline.com/${tid}/v2`,
        `https://login.microsoftonline.com/${tid}/v2.0`,
      ];
      if (!validIssuers.includes(iss)) return false;

      return true;
    } catch {
      return false;
    }
  }

  private validateRequiredFields(activity: CoreActivity): boolean {
    return !!activity.type && !!activity.serviceUrl && !!activity.conversation?.id;
  }

  private validateServiceUrl(serviceUrl: string): boolean {
    try {
      const url = new URL(serviceUrl);
      const host = url.host.toLowerCase();

      if (host === 'localhost' || host === '127.0.0.1') {
        return url.protocol === 'http:' || url.protocol === 'https:';
      }

      const allowedPatterns = [
        'botframework.com',
        'botframework.us',
        'botframework.cn',
        'trafficmanager.net',
      ];

      return allowedPatterns.some((p) => host.endsWith(p)) && url.protocol === 'https:';
    } catch {
      return false;
    }
  }

  private normalizeServiceUrl(serviceUrl: string): string {
    serviceUrl = serviceUrl.trimEnd('/');
    if (!serviceUrl.endsWith('/')) serviceUrl += '/';
    return serviceUrl;
  }

  private truncateConversationId(conversationId: string): string {
    const semicolonIndex = conversationId.indexOf(';');
    if (semicolonIndex > 0) {
      conversationId = conversationId.slice(0, semicolonIndex);
    }
    return encodeURIComponent(conversationId);
  }
}

export class BotHandlerError extends Error {
  cause: unknown;
  activity: CoreActivity;

  constructor(cause: unknown, activity: CoreActivity) {
    super('Handler error');
    this.name = 'BotHandlerError';
    this.cause = cause;
    this.activity = activity;
  }
}